### PR TITLE
refactor(share): Consolidate share pkg

### PR DESCRIPTION
**Pull request by walldiss**
Resolves https://github.com/celestiaorg/celestia-node/issues/878

### CODE REVIEW

File: cmd/flags_misc.go [hunk 1]
```diff
@@ -144,8 +144,14 @@ func ParseMiscFlags(ctx context.Context, cmd *cobra.Command) (context.Context, e
...
-			err := http.ListenAndServe("0.0.0.0:6000", mux) //nolint:gosec
-			if err != nil {
+			srv := http.Server{
+				Addr:         "0.0.0.0:6000",
+				Handler:      mux,
+				ReadTimeout:  5 * time.Second,
+				WriteTimeout: 10 * time.Second,
+			}
+
+			if err := srv.ListenAndServe(); err != nil {
```
Comment (Wondertan) [inline cmd/flags_misc.go:154]: Orthogonal

File: share/service/service.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	"github.com/ipfs/go-cid"
+
+	"github.com/celestiaorg/celestia-node/share/ipld"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+// Share is a fixed-size data chunk associated with a namespace ID, whose data will be erasure-coded and committed
+// to in Namespace Merkle trees.
+type Share = share.Share
```
Comment (Wondertan) [inline share/service/service.go:23]: This alias is now unnecessary and can be removed in favour of `share.Share`

File: share/retriever/retriever.go [hunk 3]
```diff
@@ -19,11 +22,11 @@ import (
 	"go.opentelemetry.io/otel/codes"
 	"go.opentelemetry.io/otel/trace"
 
-	"github.com/celestiaorg/celestia-node/ipld/plugin"
 	"github.com/celestiaorg/nmt"
 	"github.com/celestiaorg/rsmt2d"
 )
 
+// TODO: fix tracer and loggegr everywhere
```
Comment (Wondertan) [inline share/retriever/retriever.go:29]: typo in logger, but what does this mean?

File: share/nmt_adder.go [hunk 4]
```diff
@@ -7,26 +7,26 @@ import (
 	"github.com/ipfs/go-merkledag"
 
 	"github.com/ipfs/go-cid"
-	ipld "github.com/ipfs/go-ipld-format"
+	format "github.com/ipfs/go-ipld-format"
```
Comment (Wondertan) [inline share/nmt_adder.go:10]: If we move this file to `ipld`(main review message) then we can keep this mport as `ipld`

File: share/get_shares.go [hunk 5]
```diff
@@ -1,16 +1,16 @@
-package ipld
+package share
 
 import (
 	"context"
 	"sync"
 
+	"github.com/celestiaorg/celestia-node/share/ipld"
```
Comment (Wondertan) [inline share/get_shares.go:7]: Local imports should come after dependencies. Goland can be configured to do this automatically

Comment (Wondertan) [reply share/get_shares.go:7]: Pls check in other files as well

File: share/get.go [hunk 6]
```diff
@@ -28,9 +28,14 @@ func GetShare(
 
 // GetLeaf fetches and returns the raw leaf.
 // It walks down the IPLD NMT tree until it finds the requested one.
-func GetLeaf(ctx context.Context, bGetter blockservice.BlockGetter, root cid.Cid, leaf, total int) (ipld.Node, error) {
+func GetLeaf(
```
Comment (Wondertan) [inline share/get.go:31]: Let's move all the funcs return `format.Node` to the `ipld`. All the higher-order functions in `share` will just convert nodes/leaf into the actual shares defined in `share`.

File: share/empty.go [hunk 7]
```diff
@@ -6,10 +6,10 @@ import (
 
 	"github.com/ipfs/go-blockservice"
 	"github.com/tendermint/tendermint/pkg/consts"
-
-	"github.com/celestiaorg/celestia-node/ipld"
 )
 
+//TODO: remove?
+
```
Comment (Wondertan) [inline share/empty.go:12]: We may potentially move this to the node's share module or into the `ipld` subpkg. At first, I saw this `Ensure..` check as application specific and thus to the node's module, but then I realized that this file is largely `ipld` related and the `ipld` also has imports to app's consts like this file. Overall, it's still unclear to me what exactly to do.

File: share/availability/test/testing.go [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+	eds, err := share.AddShares(context.TODO(), shares, bServ)
+	require.NoError(t, err)
+	dah := da.NewDataAvailabilityHeader(eds)
+	return &dah
+}
+
+// RandShares provides 'n' randomized shares prefixed with random namespaces.
+func RandShares(t *testing.T, n int) []share.Share {
+	return share.RandShares(t, n)
+}
```
Comment (Wondertan) [inline share/availability/test/testing.go:44]: We may remove this now

File: share/availability/light/testing.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+// randLightService provides an unfilled share.Service with corresponding
+// blockservice.BlockService than can be filled by the test.
+func randLightService() (*service.ShareService, blockservice.BlockService) {
+	bServ := mdutils.Bserv()
+	return service.NewShareService(bServ, TestLightAvailability(bServ)), bServ
+}
+
+// randLightNode creates a Light Node filled with a random block of the given size.
+func randLightNode(dn *availability_test.DagNet, squareSize int) (*availability_test.Node, *share.Root) {
```
Comment (Wondertan) [inline share/availability/light/testing.go:35]: Public please

File: share/availability/cache/testing.go [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+	ca := NewShareAvailability(
+		light.TestLightAvailability(bServ),
+		store,
+	)
+	return service.NewShareService(bServ, ca), availability_test.RandFillBS(t, n, bServ)
+}
+
+// randFullLocalServiceWithSquare is the same as randFullServiceWithSquare, except
+// the share.Availability is wrapped with cache availability.
+func randFullLocalServiceWithSquare(t *testing.T, n int) (*service.ShareService, *share.Root) {
```
Comment (Wondertan) [inline share/availability/cache/testing.go:32]: Public

**Review by Wondertan — COMMENTED** [review]
Another async review round:
* Let's rename `retriever` pkg into `eds`, so it's `eds.Retriever`
	* This is also aligned with our plans in #1120 
* We should move `NmtNodeAdder` to `ipld`, as its primary goal is to add NMT nodes into IPLD related services
* Similarly, let's move all the funcs returning `format.Node` to the `ipld` pkg.

and needs rebase

File: share/service/service.go (hunk 2 above)
Comment (walldiss) [reply share/service/service.go]: Ty, removed it.

File: share/retriever/retriever.go (hunk 3 above)
Comment (walldiss) [reply share/retriever/retriever.go]: gone, was left from WIP

File: share/get_shares.go (hunk 5 above)
Comment (walldiss) [reply share/get_shares.go]: Thank you. I was using MakeFile and turned out there was a mistake that made -local flag to not work. Fixed

File: share/get.go (hunk 6 above)
Comment (walldiss) [reply share/get.go]: Good idea. I've splitted `GetShares` for that purpose and moved all logic related to `format.Node ` to `ipld` pkg

File: share/availability/light/testing.go (hunk 9 above)
Comment (walldiss) [reply share/availability/light/testing.go]: Yeah, might be reused somewhere else eventually. Made it public

**Review by distractedm1nd — CHANGES_REQUESTED** [review]
I really like this new structure. It is much easier to understand this part of the node now

Some other feedback:
- The imports in most files are incorrect, didn’t leave comments as to not spam
- I dont know if discovery should be in the availability subdirectory? But im not sure
- I dont think service.go needs its own package
- doc.go needs to be updated

File: share/availability/light/sample.go [hunk 11]
```diff
@@ -1,11 +1,14 @@
...
-package share
+package light
 
 import (
 	crand "crypto/rand"
 	"math/big"
 )
 
+// DefaultSampleAmount sets the default amount of samples to be sampled from the network by lightAvailability.
+var DefaultSampleAmount = 16
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply share/availability/light/sample.go:10]: It should be

[the remark this answers is not in the document]
Comment (Wondertan) [reply share/availability/light/sample.go:10]: We cannot make it a const, as it's gets overwritten

File: share/ipld/helpers.go [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+	ctx context.Context
+}
+
+func MaxSizeBatchOption(size int) ipld.BatchOption {
+	return ipld.MaxSizeBatchOption(size)
+}
+
+// IsNotFound returns if the given error is or wraps an ErrNotFound
+// (equivalent to errors.Is(err, ErrNotFound{}))
+func IsNotFound(err error) bool {
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply share/ipld/helpers.go:26]: Yeah let's remove it

**Comment by Wondertan** [comment]
Will answer some while I am at it:
* Discover as part of availability - yes
* Service _implementation_ as it's own pkg solves cyclic import. Didn't see a better solution. We were also thinking about moving service implementation and definition to the share module. Share library does not have to have it and can only provide primitives to builds it

File: share/availability.go [hunk 13]
```diff
@@ -23,3 +33,13 @@ type Availability interface {
 	// TODO(@Wondertan): Merge with SharesAvailable method, eventually
 	ProbabilityOfAvailability() float64
 }
+
+// translate transforms square coordinates into IPLD NMT tree path to a leaf node.
+// It also adds randomization to evenly spread fetching from Rows and Columns.
+func Translate(dah *Root, row, col int) (cid.Cid, int) {
```
[the remark this answers is not in the document]
Comment (walldiss) [reply share/availability.go:39]: I've placed `Root` here because it is the only place it is used in the `share` pkg. If there is a need to get info about `Root`, it is always in context of `share.Avalability`, that is residing nearby. I feel like having it grouped is increasing variable locality and reducing cognitive load.
 On the other hand `Share` is more widely used and would be nice to have it more atomic.

File: share/availability/full/availability.go [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+// to download enough shares to fully reconstruct the data square.
+type ShareAvailability struct {
+	rtrv *eds.Retriever
+	disc *discovery.Discovery
+
+	cancel context.CancelFunc
+}
+
+// NewShareAvailability creates a new full ShareAvailability.
+func NewShareAvailability(bServ blockservice.BlockService, disc *discovery.Discovery) *ShareAvailability {
```
[the remark this answers is not in the document]
Comment (walldiss) [reply share/availability/full/availability.go:32]: Thank you, somehow `NewLightAvailability` got missed, I've fixed it to be consistent. The idea was to to leave type of availability to pkg name. So caller could call `full.NewShareAvailability` or `light.NewShareAvailability`. Since they all are implementation of same `share.Availability` interface.

File: share/ipld/helpers.go [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+import (
+	"fmt"
+)
+
+func SanityCheckNID(nID []byte) error {
+	if len(nID) != NamespaceSize {
+		return fmt.Errorf("expected namespace ID of size %d, got %d", NamespaceSize, len(nID))
+	}
+	return nil
+}
```
Comment (Wondertan) [inline share/ipld/helpers.go:12]: It doesn't worth much if it has its own file. Also, ipld pkg is not the place it should be in anyway. Can you just inline it inside the service and remove it completely from GetSharesByNamespace. The service anyway does the check before passing it down to GetSharesByNamespsace.

File: share/ipld/nmt.go [hunk 16]
```diff
@@ -26,11 +33,22 @@ const (
 
 	// Sha256Namespace8Flagged is the multihash code used to hash blocks
 	// that contain an NMT node (inner and leaf nodes).
-	sha256Namespace8Flagged = 0x7701
+	Sha256Namespace8Flagged = 0x7701
```
Comment (Wondertan) [inline share/ipld/nmt.go:36]: Why public?

**Review by Wondertan — COMMENTED** [review]
Last nit, other than that LGTM. 

We should also do something about changes unrelated to the PR: makefile and HTTP server

File: share/availability.go [hunk 17]
```diff
@@ -3,7 +3,13 @@ package share
...
+	"math/rand"
 	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/tendermint/tendermint/pkg/da"
+
+	"github.com/celestiaorg/celestia-node/share/ipld"
 )
 
 // ErrNotAvailable is returned whenever DA sampling fails.
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply share/availability.go:15]: (The error below the comment )

File: share/ipld/helpers.go (hunk 15 above)
Comment (walldiss) [reply share/ipld/helpers.go]: Good point, thank you. File is gone.

File: share/share.go [hunk 18]
```diff
@@ -1,164 +1,106 @@
...
 
-// NewService creates a new basic share.Module.
-func NewService(bServ blockservice.BlockService, avail Availability) *Service {
-	return &Service{
-		rtrv:         ipld.NewRetriever(bServ),
-		Availability: avail,
-		bServ:        bServ,
-	}
+// ShareData gets data from the share.
+func ShareData(s Share) []byte { //nolint:revive
```
[the remark this answers is not in the document]
Comment (walldiss) [reply share/share.go]: removed share prefixes

File: share/ipld/nmt.go (hunk 16 above)
Comment (walldiss) [reply share/ipld/nmt.go]: Should be private now

File: share/availability.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (Wondertan) [reply share/availability.go]: Move to ipld now

File: share/get.go [hunk 19]
```diff
@@ -93,7 +107,7 @@ func GetProof(
 	leaf, total int,
 ) ([]cid.Cid, error) {
 	// request the node
-	nd, err := plugin.GetNode(ctx, bGetter, root)
+	nd, err := ipld.GetNode(ctx, bGetter, root)
```
Comment (Wondertan) [inline share/get.go:110]: This GetProofs also deserves to be moved in `ipld`

Comment (walldiss) [reply share/get.go]: good point, move it to same file where GetLeaf is.

File: share/availability.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (walldiss) [reply share/availability.go:11]: renamed in previous commit

File: share/availability.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (walldiss) [reply share/availability.go]: moved to `ipld` pkg

File: Makefile [hunk 20]
```diff
@@ -67,7 +67,7 @@ install-key:
 ## fmt: Formats only *.go (excluding *.pb.go *pb_test.go). Runs `gofmt & goimports` internally.
 fmt:
 	@find . -name '*.go' -type f -not -path "*.git*" -not -name '*.pb.go' -not -name '*pb_test.go' | xargs gofmt -w -s
-	@find . -name '*.go' -type f -not -path "*.git*"  -not -name '*.pb.go' -not -name '*pb_test.go' | xargs goimports -w -local github.com/celestiaorg
+	@find . -name '*.go' -type f -not -path "*.git*"  -not -name '*.pb.go' -not -name '*pb_test.go' | xargs goimports -w -local "github.com/celestiaorg"
```
Comment (Wondertan) [inline Makefile:70]: Pls extract to a separate PR

File: cmd/flags_misc.go [hunk 21]
```diff
@@ -139,8 +140,14 @@ func ParseMiscFlags(ctx context.Context, cmd *cobra.Command) (context.Context, e
...
```
Comment (Wondertan) [inline cmd/flags_misc.go:150]: Pls extract as well

**Review by Wondertan — DISMISSED** [review]
Approving all the changes here, and they are awesome! One last ask. Let's exclude all the unrelated changes and make it a habit

**Review by Wondertan — DISMISSED** [review]

**Review by distractedm1nd — DISMISSED** [review]

**Review by Wondertan — APPROVED** [review]

File: nodebuilder/share/service.go [hunk 22]
```diff
@@ -30,14 +32,14 @@ type Module interface {
 }
 
 func NewModule(lc fx.Lifecycle, bServ blockservice.BlockService, avail share.Availability) Module {
-	service := share.NewService(bServ, avail)
+	svc := service.NewShareService(bServ, avail)
```
Comment (renaynay) [inline nodebuilder/share/service.go:35]: small nit but we usually refer to service as `serv`

File: share/availability/cache/testing.go [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+	mdutils "github.com/ipfs/go-merkledag/test"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/availability/full"
+	"github.com/celestiaorg/celestia-node/share/availability/light"
+	availability_test "github.com/celestiaorg/celestia-node/share/availability/test"
+	"github.com/celestiaorg/celestia-node/share/service"
+)
+
+// RandLightLocalServiceWithSquare is the same as randLightServiceWithSquare, except
```
Comment (renaynay) [inline share/availability/cache/testing.go:17]: `randLightServiceWithSquare` doesn't exist in this pkg anymore, same with `randFullServiceWithSquare`

File: share/availability/full/reconstruction_test.go [hunk 24]
```diff
@@ -10,19 +10,22 @@ import (
...
+	"github.com/celestiaorg/celestia-node/share/eds"
 )
 
 func init() {
-	ipld.RetrieveQuadrantTimeout = time.Millisecond * 100 // to speed up tests
+	eds.RetrieveQuadrantTimeout = time.Millisecond * 100 // to speed up tests
 }
 
-// TestShareAvailable_OneFullNode asserts that a FullAvailability node can ensure
+// TestShareAvailable_OneFullNode asserts that a Full node can ensure
```
Comment (renaynay) [inline share/availability/full/reconstruction_test.go:23]:
```suggestion
// TestShareAvailable_OneFullNode asserts that a full node can ensure
```
nit but we capitalise some and not others

**Review by renaynay — COMMENTED** [review]
Some import formatting issues and other nits but looks good!

File: das/daser_test.go [hunk 25]
```diff
@@ -29,7 +32,7 @@ var timeout = time.Second * 15
 func TestDASerLifecycle(t *testing.T) {
 	ds := ds_sync.MutexWrap(datastore.NewMapDatastore())
 	bServ := mdutils.Bserv()
-	avail := share.TestLightAvailability(bServ)
+	avail := light.TestLightAvailability(bServ)
```
[the remark this answers is not in the document]
Comment (walldiss) [reply das/daser_test.go]: renamed

File: fraud/bad_encoding.go [hunk 26]
```diff
@@ -5,16 +5,18 @@ import (
 	"errors"
 	"fmt"
 
+	"github.com/celestiaorg/celestia-node/share/ipld"
```
[the remark this answers is not in the document]
Comment (walldiss) [reply fraud/bad_encoding.go]: fixed

File: fraud/bad_encoding_test.go [hunk 27]
```diff
@@ -6,10 +6,10 @@ import (
 	"testing"
 	"time"
 
+	"github.com/celestiaorg/celestia-node/share"
```
[the remark this answers is not in the document]
Comment (walldiss) [reply fraud/bad_encoding_test.go]: and this one

File: nodebuilder/share/service.go (hunk 22 above)
Comment (walldiss) [reply nodebuilder/share/service.go]: renamed for consistency.

File: share/availability/cache/testing.go (hunk 23 above)
Comment (walldiss) [reply share/availability/cache/testing.go]: fixed to reference `light.RandServiceWithSquare` instead

File: share/availability/discovery/discovery.go [hunk 28]
```diff
@@ -1,16 +1,19 @@
...
 
+	logging "github.com/ipfs/go-log/v2"
 	core "github.com/libp2p/go-libp2p-core/discovery"
 	"github.com/libp2p/go-libp2p-core/event"
 	"github.com/libp2p/go-libp2p-core/host"
 	"github.com/libp2p/go-libp2p-core/network"
 	"github.com/libp2p/go-libp2p-core/peer"
 )
 
+var log = logging.Logger("share/discovery")
```
[the remark this answers is not in the document]
Comment (walldiss) [reply share/availability/discovery/discovery.go]: fixed both

[the remark this answers is not in the document]
Comment (Wondertan) [reply share/availability/discovery/discovery.go]: The problem with this naming is that it's too long and the logger cuts it out

[the remark this answers is not in the document]
Comment (renaynay) [reply share/availability/discovery/discovery.go:15]: good with me. Also thought it might be too verbose

**Review by renaynay — APPROVED** [review]

**Review by Wondertan — APPROVED** [review]
